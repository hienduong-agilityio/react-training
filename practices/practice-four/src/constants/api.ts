export const HTTP_METHODS = {
  GET: 'GET',
  POST: 'POST',
  PUT: 'PUT',
  DELETE: 'DELETE'
};

export const ROOT_API = import.meta.env.VITE_API_ENDPOINT;

export const API_URL = {
  PROJECT: `${ROOT_API}/projects`
};
