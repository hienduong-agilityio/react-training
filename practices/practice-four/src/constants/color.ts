// Consider using enums instead of constants
export const COLORS_CLASS = {
  DEFAULT: 'color-default',
  PRIMARY: 'color-primary',
  SUCCESS: 'color-success',
  WARNING: 'color-warning',
  DANGER: 'color-danger '
};
