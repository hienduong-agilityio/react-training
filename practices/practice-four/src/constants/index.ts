// Index constants files
export * from './api';
export * from './color';
export * from './filerSearchOption';
export * from './route';
export * from './rowPerPage';
export * from './store';
export * from './validationMessages';
