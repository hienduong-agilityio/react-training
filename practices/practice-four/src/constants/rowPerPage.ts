export const ROWS_PER_PAGE_OPTIONS = [
  { value: 5, label: '5' },
  { value: 10, label: '10' },
  { value: 15, label: '15' },
  { value: 20, label: '20' },
  { value: 30, label: '30' },
  { value: 40, label: '40' },
  { value: 50, label: '50' }
];
