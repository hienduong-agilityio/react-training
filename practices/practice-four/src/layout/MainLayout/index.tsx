// Libraries
import { Outlet } from 'react-router-dom';

// Layouts
import { Sidebar, Navbar, TabLayout } from '@/layout';

/**
 * MainLayout component
 *
 * @returns {JSX.Element} The main layout structure with navigation components and content area.
 */
export const MainLayout = (): JSX.Element => {
  return (
    <main className='flex'>
      <Sidebar />
      <div className='flex flex-col w-full'>
        <Navbar />
        <TabLayout />
        <Outlet />
      </div>
    </main>
  );
};
