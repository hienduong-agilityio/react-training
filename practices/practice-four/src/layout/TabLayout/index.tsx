// Libraries
import { generateBreadcrumbItems } from '@/helpers/breadcrumb';
import { useMemo } from 'react';
import { useLocation, useParams } from 'react-router-dom';

// Components
import { Breadcrumbs } from '@/components';

/**
 * TabLayout component
 *
 * @returns {JSX.Element} - The TabLayout element
 */
export const TabLayout = (): JSX.Element => {
  const location = useLocation();
  const { id: projectId } = useParams();

  const breadcrumbItems = useMemo(
    () => generateBreadcrumbItems(location.pathname, projectId),
    [location.pathname, projectId]
  );

  return (
    <div className='px-4 py-2'>
      <Breadcrumbs items={breadcrumbItems} />
    </div>
  );
};
