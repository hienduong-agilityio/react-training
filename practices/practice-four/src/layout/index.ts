// Index Layout files
export * from './MainLayout';
export * from './NavBar';
export * from './ProjectLayout';
export * from './SideBar';
export * from './TabLayout';
