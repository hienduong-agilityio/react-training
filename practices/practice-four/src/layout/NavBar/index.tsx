// Libraries
import { Link } from 'react-router-dom';

// Components
import { Avatar } from '@/components';

// Svg
import notifyIcon from '@public/images/notificationIcon.svg';
import helpIcon from '@public/images/helpIcon.svg';

// Constants
import { ROUTE } from '@/constants';

/**
 * Navbar components
 *
 * @returns {JSX.Element} The Navbar component.
 */
export const Navbar = (): JSX.Element => {
  return (
    <nav className='bg-light w-full border'>
      <div className='flex flex-wrap items-center justify-between p-4'>
        <Link to={ROUTE.PROJECT} className='text-2xl font-semibold text-gray-900'>
          Projects
        </Link>
        <div className='flex items-center gap-6'>
          <img src={notifyIcon} className='cursor-not-allowed' alt='Notification' />
          <img src={helpIcon} className='cursor-not-allowed' alt='Help' />
          <Avatar customClass='rounded-full cursor-not-allowed' name='HienDuong' />
        </div>
      </div>
    </nav>
  );
};
