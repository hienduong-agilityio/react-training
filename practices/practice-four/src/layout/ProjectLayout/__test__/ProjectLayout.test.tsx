// Libraries
import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';

// Components
import { ProjectLayout } from '@/layout';

// Constants
import { ROUTE } from '@/constants';

// Mock the images
jest.mock('@public/images/backArrow.svg', () => 'mock-back-arrow');

// Mock useNavigate
const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate
}));

/**
 ProjectLayout
  √ should render the layout and back button
  √ should navigate back if history exists
  √ should navigate to the project route if no history exists
  √ should render the Outlet for child components
  √ should match the snapshot
*/
describe('ProjectLayout', () => {
  const renderProjectLayout = (initialEntries = ['/project-layout']) => {
    return render(
      <MemoryRouter initialEntries={initialEntries}>
        <Routes>
          <Route path='/' element={<div>Home Page</div>} />
          <Route path={ROUTE.PROJECT} element={<div>Projects Page</div>} />
          {/* Adding a route that uses ProjectLayout */}
          <Route path='/project-layout' element={<ProjectLayout />}>
            <Route path='' element={<div>Child Component</div>} />
          </Route>
        </Routes>
      </MemoryRouter>
    );
  };

  it('should match the snapshot', () => {
    const { asFragment } = renderProjectLayout();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the layout and back button', async () => {
    renderProjectLayout();
    const backButton = await screen.findByRole('button');
    expect(backButton).toBeInTheDocument();
    expect(screen.getByAltText('Back Arrow')).toHaveAttribute('src', 'mock-back-arrow');
  });

  it('should navigate back if history exists', async () => {
    window.history.pushState({}, 'Test Page', '/project-layout');
    renderProjectLayout(['/project-layout']);

    const backButton = await screen.findByRole('button');
    fireEvent.click(backButton);

    expect(mockNavigate).toHaveBeenCalledWith(-1);
  });

  it('should navigate to the project route if no history exists', async () => {
    window.history.replaceState({}, 'Test Page', '/');

    Object.defineProperty(window, 'history', {
      writable: true,
      value: { length: 1 }
    });

    renderProjectLayout(['/project-layout']);

    const backButton = await screen.findByRole('button');
    fireEvent.click(backButton);

    expect(mockNavigate).toHaveBeenCalledWith(ROUTE.PROJECT);
  });

  it('should render the Outlet for child components', async () => {
    renderProjectLayout(['/project-layout']);
    expect(await screen.findByText('Child Component')).toBeInTheDocument();
  });
});
