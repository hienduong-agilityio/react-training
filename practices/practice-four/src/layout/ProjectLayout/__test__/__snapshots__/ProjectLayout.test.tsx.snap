// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectLayout should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col h-full bg-gray-300 p-6"
  >
    <button
      class="mb-6 w-24 p-2 bg-white border-2 text-black flex gap-2 items-center justify-center focus:ring-4 focus:ring-primary-100 rounded-md cursor-pointer"
    >
      <img
        alt="Back Arrow"
        class="w-5 h-5 pr-2"
        src="mock-back-arrow"
      />
       Back
    </button>
    <div
      class="bg-white p-6 rounded-lg shadow-md w-full max-w h-auto"
    >
      <div>
        Child Component
      </div>
    </div>
  </div>
</DocumentFragment>
`;
